app/src/main/java/org/wikipedia/edit/SyntaxHighlightViewAdapter.kt

```diff
@@ -0,0 +1,131 @@
+    val wikiSite: WikiSite,
+    private val rootView: View,
+    val editText: SyntaxHighlightableEditText,
+    private val wikiTextKeyboardView: WikiTextKeyboardView,
+    private val wikiTextKeyboardFormattingView: WikiTextKeyboardFormattingView,
+    private val wikiTextKeyboardHeadingsView: WikiTextKeyboardHeadingsView,
+    showUserMention: Boolean = false
+) : WikiTextKeyboardView.Callback {
+
+    init {
+        wikiTextKeyboardView.editText = editText
+        wikiTextKeyboardView.callback = this
+        wikiTextKeyboardFormattingView.editText = editText
+        wikiTextKeyboardFormattingView.callback = this
+        wikiTextKeyboardHeadingsView.editText = editText
+        wikiTextKeyboardHeadingsView.callback = this
+        wikiTextKeyboardView.userMentionVisible = showUserMention
+
+        activity.window.decorView.viewTreeObserver.addOnGlobalLayoutListener {
+            activity.window.decorView.post {
+                if (!activity.isDestroyed) {
+                    showOrHideSyntax(editText.hasFocus())
+                }
+            }
+        }
+
+        editText.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
+            showOrHideSyntax(hasFocus)
+        }
+    }
+
+    private val requestLinkFromSearch = activity.registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+        if (it.resultCode == SearchActivity.RESULT_LINK_SUCCESS) {
+            it.data?.parcelableExtra<PageTitle>(SearchActivity.EXTRA_RETURN_LINK_TITLE)?.let { title ->
+                wikiTextKeyboardView.insertLink(title, wikiSite.languageCode)
+            }
+        }
+    }
+
+    private val requestInsertMedia = activity.registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+        if (it.resultCode == InsertMediaActivity.RESULT_INSERT_MEDIA_SUCCESS) {
+            it.data?.let { intent ->
+                editText.inputConnection?.commitText("${intent.getStringExtra(
+                    InsertMediaActivity.RESULT_WIKITEXT)}", 1)
+            }
+        }
+    }
+
+    override fun onPreviewLink(title: String) {
+        val dialog = LinkPreviewDialog.newInstance(HistoryEntry(PageTitle(title, wikiSite), HistoryEntry.SOURCE_INTERNAL_LINK), null)
+        ExclusiveBottomSheetPresenter.show(activity.supportFragmentManager, dialog)
+        editText.post {
+            dialog.dialog?.setOnDismissListener {
+                if (!activity.isDestroyed) {
+                    editText.postDelayed({
+                        DeviceUtil.showSoftKeyboard(editText)
+                    }, 200)
+                }
+            }
+        }
+    }
+
+    override fun onRequestInsertMedia() {
+        requestInsertMedia.launch(InsertMediaActivity.newIntent(activity, wikiSite, ""))
+    }
+
+    override fun onRequestInsertLink() {
+        requestLinkFromSearch.launch(SearchActivity.newIntent(activity, Constants.InvokeSource.TALK_REPLY_ACTIVITY, null, true))
```

Comment:
Do we need to add a parameter to the adapter in order to send the correct source here?

Reply:
yes!